lucille-plugins/lucille-video/src/main/java/com/kmwllc/lucille/video/VideoFileHandler.java

```diff
@@ -0,0 +1,274 @@
+      this.frameImageField = frameImageField;
+    }
+
+    private void closeAll() {
+      if (closed) {
+        return;
+      }
+      closed = true;
+
+      try {
+        grabber.stop();
+      } catch (Exception ignored) {}
+
+      try {
+        grabber.release();
+      } catch (Exception ignored) {}
+    }
+
+    @Override
+    public boolean hasNext() {
+      if (closed || nextDoc != null) {
+        return nextDoc != null;
+      }
+
+      if (!started) {
+        try {
+          grabber.start();
+          started = true;
+        } catch (Exception e) {
+          closeAll();
+          throw new RuntimeException("Unable to start FFmpegFrameGrabber for: " + pathStr, e);
+        }
+      }
+
+      try {
+        while (true) {
+          Frame frame = grabber.grabImage();
+
+          if (frame == null) {
+            closeAll();
+            return false;
+          }
+
+          int index = videoFrameIndex++;
+          if (frameStride > 1 && (index % frameStride) != 0) {
+            continue;
+          }
+
+          BufferedImage image = converter.convert(frame);
+          if (image == null) {
+            continue;
+          }
+
+          long timeUs = grabber.getTimestamp();
+          long timeMs = timeUs / 1000L;
+
+          byte[] pngBytes = null;
+          try (ByteArrayOutputStream stream = new ByteArrayOutputStream()) {
+            ImageIO.write(image, "png", stream);
+            stream.flush();
+            pngBytes = stream.toByteArray();
+          } catch (IOException e) {
+            log.warn("PNG encoding failed for {} frame {}: {}", pathStr, index, e.toString());
+          }
+
+          String frameId = parentId + "-f" + index;
+          Document doc = Document.create(frameId);
+          doc.setField(sourceField, pathStr);
+          doc.setField(fileNameField, fileName);
+          doc.setField(frameIndexField, index);
+          doc.setField(frameTimeMsField, timeMs);
+          doc.setField(frameTimecodeField, formatTimecode(timeMs));
+          doc.setField(imageWidthField, image.getWidth());
+          doc.setField(imageHeightField, image.getHeight());
+          doc.setField(frameImageField, pngBytes);
+
+          nextDoc = doc;
+          return true;
+        }
+      } catch (Exception e) {
+        closeAll();
+        log.error("Error while grabbing frames for {}", pathStr, e);
+        return false;
+      }
+    }
+
+    @Override
+    public Document next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+      Document out = nextDoc;
+      nextDoc = null;
+      return out;
+    }
+  }
+
+  private String buildParentId(String pathStr) {
+    String leaf = FilenameUtils.getName(pathStr);
+    String stem = FilenameUtils.removeExtension(leaf);
+    String raw = (docIdFormat != null) ? String.format(docIdFormat, stem) : stem;
+    return docIdPrefix + raw;
+  }
+
+  private static String formatTimecode(long millis) {
```

Comment:
should have some test coverage that shows that frameTimeCodeField is getting populated with the expected string... implying that this formatTimecode method is generating the right output